extensions/content-type-builder/services/relation-builder/relation-template.js

```diff
@@ -1,33 +1,91 @@
 const { sanitizeEntity } = require('strapi-utils');
 
-module.exports = {
-  async createRelation(ctx) {
-    const { relationName } = ctx.params; 
-    let data;
-    if (ctx.query._q) {
-      data = await strapi.services.placeholder.search(ctx.query);
-    } else {
-      data  = await strapi.services.placeholder.find(ctx.query);
-    }
-    const copiedData = data.map(el => ({ ...el }));
-    copiedData.forEach(el => {
-      let parents = copiedData.filter(x => {
-        if(x[relationName]) {
-          return x[relationName].id === el.id
-        } else {
-          return false
-        }
-      })
-      if(parents.length) {
-        parents.forEach( parent => {
-          if(parent.id !== el.id) {
-            parent.child = el;
-            el.isChild = true;
-          }
-        })
+const createRelationStructure = (data, relationName) => {
+  const copiedData = data.map(el => ({ ...el }));
+  copiedData.forEach(el => {
+    let parents = copiedData.filter(x => {
+      if(x[relationName]) {
+        return x[relationName].id === el.id
+      } else {
+        return false
       }
     })
-    return copiedData.filter(el => !el.isChild)
-  },
-};
+    if(parents.length) {
+      parents.forEach( parent => {
+        if(parent.id !== el.id) {
+          parent.child = el;
+          el.isChild = true;
+        }
+      })
+    }
+  })
+  return copiedData.filter(el => !el.isChild)
+}
+
+const getIdFromParentAndChild = (obj, idsContainter, relationName) => {
+  if(idsContainter.includes(obj.id)) {
+    idsContainter.push(obj.id);
+    return;
+  }
+  idsContainter.push(obj.id);
+  if(obj.child) {
+    getIdFromParentAndChild(obj.child, idsContainter, relationName);
+  }
+  if(obj[relationName] && obj[relationName].id === obj.id){
+    idsContainter.push(obj.id)
+  }
+}
+
+const prepareNewStructure = (childId, parentId,  data, relationName) => {
+  let parentToChangeAChild = data.find(el => el.id == parentId);
+  let newChild;
+  if(childId == -1) {
+    newChild = null;
+  } else {
+    newChild = data.find(el => el.id == childId);
+  }
+  parentToChangeAChild[relationName] = newChild;
+  const updatedData  = data.map(el => el.id == parentId ? parentToChangeAChild : el);
+  return createRelationStructure(updatedData, relationName)
+}
+
+const collectIdsFromNestedData = (nestedData, relationName) => {
+  let ids = [];
+  nestedData.forEach(el => {
+    getIdFromParentAndChild(el, ids, relationName)
```

Comment:
map would be better and cleaner, instead of allow to mutate parameters inside function (`getIdFromParentAndChild` will add something to one of their parameters - `ids`)